components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java

```diff
@@ -0,0 +1,952 @@
+    }
+
+    /**
+     * Evaluates if the client secret should be generated and if required, generate and store the secret.
+     *
+     * @param context                   Authentication context.
+     * @param authenticatorProperties   Authenticator properties.
+     * @throws AuthenticationFailedException If an error occurs when generating the client secret.
+     */
+    private void evaluateClientSecret(AuthenticationContext context, Map<String, String> authenticatorProperties)
+            throws AuthenticationFailedException {
+
+        String tenantDomain = context.getTenantDomain();
+        long currentEpochTime = Instant.now().getEpochSecond();
+        long expiryEpochTime = 0;
+
+        try {
+            if (StringUtils.isNotBlank(authenticatorProperties.get(
+                    AppleAuthenticatorConstants.CLIENT_SECRET_EXPIRY_TIME_METADATA))) {
+                expiryEpochTime = Long.parseLong(authenticatorProperties.get(AppleAuthenticatorConstants
+                        .CLIENT_SECRET_EXPIRY_TIME_METADATA)) - AppleAuthenticatorConstants
+                        .CLIENT_SECRET_VALIDITY_PERIOD_THRESHOLD;
+            }
+        } catch (NumberFormatException e) {
+            throw new AuthenticationFailedException(
+                    AppleErrorConstants.ErrorMessages.UNSUPPORTED_VALUE_PROVIDED_FOR_AUTHENTICATOR_PROPERTY.getCode(),
+                    String.format(AppleErrorConstants.ErrorMessages
+                            .UNSUPPORTED_VALUE_PROVIDED_FOR_AUTHENTICATOR_PROPERTY.getMessage(),
+                            AppleAuthenticatorConstants.CLIENT_SECRET_EXPIRY_TIME_METADATA, tenantDomain)
+            );
+        }
+
+        if (StringUtils.isBlank(authenticatorProperties.get(OIDCAuthenticatorConstants.CLIENT_SECRET)) ||
+                Boolean.parseBoolean(
+                        authenticatorProperties.get(AppleAuthenticatorConstants.REGENERATE_CLIENT_SECRET)) ||
+                expiryEpochTime <= currentEpochTime
+        ) {
+            if (log.isDebugEnabled()) {
+                String msg = String.format("Generating client secret since it is null or invalidated. " +
+                        "Tenant: %s, Secret re-gen property: %s, Current time: %s, Expiry time: %s.", tenantDomain,
+                        authenticatorProperties.get(AppleAuthenticatorConstants.REGENERATE_CLIENT_SECRET),
+                        currentEpochTime, expiryEpochTime);
+                log.debug(msg);
+            }
+
+            // Check for pre-conditions.
+            if (StringUtils.isBlank(authenticatorProperties.get(AppleAuthenticatorConstants.PRIVATE_KEY))
+                    || StringUtils.isBlank(authenticatorProperties.get(AppleAuthenticatorConstants.TEAM_ID))
+                    || StringUtils.isBlank(authenticatorProperties.get(OIDCAuthenticatorConstants.CLIENT_ID))
+                    || StringUtils.isBlank(authenticatorProperties.get(AppleAuthenticatorConstants.KEY_ID))
+            ) {
```

Comment:
Shouldn't these be mandatory?

Reply:
Yes. These attributes are set as required fields in [authenticator configs](https://github.com/wso2-extensions/identity-outbound-auth-apple/pull/1/files#diff-7d561c19910ba6eb5ad3e88a271822acdd95c67bfc6dda4cb96e2a595a12c99fR343). However it is possible to set them empty through the API/ management console.
Therefore I have added a check here.